File: results/2026-06-26_103905/tasks/11-semantic-version-bumper/default-opus48-1m-high/generated-code/src/bumper.js
// Core library for the semantic version bumper.
//
// The parsing/decision helpers are pure functions; the read/write helpers touch
// the filesystem but are kept tiny and focused so they're still easy to test
// against temp files. The thin CLI wrapper in ./cli.js wires everything to
// process arguments.

import fs from "node:fs";
import path from "node:path";

// Commits in a fixture / git-dump log are separated by this delimiter line.
// (Produce one with: `git log --format='%B%n--==COMMIT==--'`.)
export const COMMIT_DELIMITER = "--==COMMIT==--";

// Conventional-commit header: type(scope)!: description
//   group 1: type        (required, e.g. feat, fix, chore)
//   group 2: scope        (optional, inside parentheses)
//   group 3: "!"          (optional breaking-change marker)
//   group 4: description  (required)
const HEADER_RE = /^([a-zA-Z]+)(?:\(([^)]*)\))?(!)?:\s*(.+)$/;

/**
 * Parse a raw commit log into structured conventional-commit objects.
 *
 * Non-conventional entries (merge commits, blank blocks, free-form messages)
 * are silently skipped so the caller only ever sees commits that can influence
 * the version bump.
 *
 * @param {string} log raw log text, commits separated by COMMIT_DELIMITER
 * @returns {Array<{type:string, scope:string|null, breaking:boolean, description:string}>}
 */
export function parseCommits(log) {
  if (typeof log !== "string") {
    throw new TypeError("parseCommits expects the commit log as a string");
  }

  const blocks = log.split(COMMIT_DELIMITER);
  const commits = [];

  for (const block of blocks) {
    const trimmed = block.trim();
    if (!trimmed) continue;

    const lines = trimmed.split("\n");
    const header = lines[0].trim();
    const body = lines.slice(1).join("\n");

    const match = HEADER_RE.exec(header);
    if (!match) continue; // not a conventional commit -> ignore

    const [, type, scope, bang, description] = match;
    const breaking =
      bang === "!" || /^BREAKING[ -]CHANGE:/m.test(body);

    commits.push({
      type: type.toLowerCase(),
      scope: scope ? scope.trim() : null,
      breaking,
      description: description.trim(),
    });
  }

  return commits;
}

// Precedence of bumps: a single breaking change forces a major; otherwise any
// feat triggers a minor; otherwise any fix triggers a patch. Everything else
// (chore, docs, style, refactor, test, ci, ...) does not move the version.
const BUMP_RANK = { major: 3, minor: 2, patch: 1 };

/**
 * Decide the bump level implied by a set of parsed commits.
 *
 * @param {Array} commits output of parseCommits
 * @returns {"major"|"minor"|"patch"|null} null when nothing warrants a bump
 */
export function determineBump(commits) {
  let best = null;
  for (const commit of commits) {
    let level = null;
    if (commit.breaking) level = "major";
    else if (commit.type === "feat") level = "minor";
    else if (commit.type === "fix") level = "patch";

    if (level && (!best || BUMP_RANK[level] > BUMP_RANK[best])) {
      best = level;
    }
  }
  return best;
}

// Strict SemVer core (major.minor.patch). Pre-release / build metadata is out
// of scope for this bumper; we keep the surface small and reject anything else.
const VERSION_RE = /^v?(\d+)\.(\d+)\.(\d+)$/;

/**
 * Apply a bump level to a version string.
 *
 * @param {string} version e.g. "1.4.2" (an optional leading "v" is tolerated)
 * @param {"major"|"minor"|"patch"} bumpType
 * @returns {string} the new version, without any leading "v"
 */
export function bumpVersion(version, bumpType) {
  const match = VERSION_RE.exec(String(version).trim());
  if (!match) {
    throw new Error(
      `Invalid semantic version: "${version}". Expected MAJOR.MINOR.PATCH (e.g. 1.2.3).`
    );
  }

  let [major, minor, patch] = match.slice(1).map(Number);

  switch (bumpType) {
    case "major":
      major += 1;
      minor = 0;
      patch = 0;
      break;
    case "minor":
      minor += 1;
      patch = 0;
      break;
    case "patch":
      patch += 1;
      break;
    default:
      throw new Error(
        `Unknown bump type: "${bumpType}". Expected "major", "minor" or "patch".`
      );
  }

  return `${major}.${minor}.${patch}`;
}

// A file is treated as a package.json when its basename is "package.json";
// anything else (VERSION, .version, version.txt, ...) is a plain version file.
function isPackageJson(filePath) {
  return path.basename(filePath) === "package.json";
}

/**
 * Read the current version from either a package.json or a plain version file.
 *
 * @param {string} filePath
 * @returns {string} the version string (e.g. "1.2.3")
 */
export function readVersion(filePath) {
  let raw;
  try {
    raw = fs.readFileSync(filePath, "utf8");
  } catch {
    throw new Error(`Version file not found: ${filePath}`);
  }

  if (isPackageJson(filePath)) {
    let pkg;
    try {
      pkg = JSON.parse(raw);
    } catch (e) {
      throw new Error(`Could not parse ${filePath} as JSON: ${e.message}`);
    }
    if (typeof pkg.version !== "string") {
      throw new Error(`${filePath} has no "version" field`);
    }
    return pkg.version.trim();
  }

  const version = raw.trim();
  if (!version) {
    throw new Error(`Version file is empty: ${filePath}`);
  }
  return version;
}

/**
 * Persist a new version back to the same file, preserving its format.
 * For package.json the surrounding keys/indentation are kept intact.
 *
 * @param {string} filePath
 * @param {string} newVersion
 */
export function writeVersion(filePath, newVersion) {
  if (isPackageJson(filePath)) {
    const pkg = JSON.parse(fs.readFileSync(filePath, "utf8"));
    pkg.version = newVersion;
    // 2-space indent + trailing newline matches npm's own conventions.
    fs.writeFileSync(filePath, JSON.stringify(pkg, null, 2) + "\n");
  } else {
    fs.writeFileSync(filePath, newVersion + "\n");
  }
}

// Conventional-commit type -> changelog section heading. Order here is the
// order sections appear in the output.
const SECTIONS = [
  { heading: "Breaking Changes", match: (c) => c.breaking },
  { heading: "Features", match: (c) => !c.breaking && c.type === "feat" },
  { heading: "Bug Fixes", match: (c) => !c.breaking && c.type === "fix" },
];

/**
 * Build a Markdown changelog entry for a release.
 *
 * @param {string} version the new version being released
 * @param {Array} commits parsed commits (output of parseCommits)
 * @param {string} dateStr ISO date (YYYY-MM-DD) for the release header
 * @returns {string} a Markdown block ending in a single trailing newline
 */
export function generateChangelog(version, commits, dateStr) {
  const lines = [`## [${version}] - ${dateStr}`, ""];

  for (const section of SECTIONS) {
    const matched = commits.filter(section.match);
    if (matched.length === 0) continue; // omit empty sections

    lines.push(`### ${section.heading}`, "");
    for (const c of matched) {
      const prefix = c.scope ? `**${c.scope}:** ` : "";
      lines.push(`- ${prefix}${c.description}`);
    }
    lines.push("");
  }

  return lines.join("\n").trimEnd() + "\n";
}

// Prepend a new changelog entry above any existing content, keeping a stable
// top-of-file title. Creates the file if it does not exist yet.
function prependChangelog(changelogFile, entry) {
  const TITLE = "# Changelog\n";
  let existingBody = "";
  if (fs.existsSync(changelogFile)) {
    let current = fs.readFileSync(changelogFile, "utf8");
    // Drop a leading "# Changelog" title if present; we re-add it ourselves.
    current = current.replace(/^# Changelog\s*\n+/, "");
    existingBody = current;
  }
  const out = `${TITLE}\n${entry}${existingBody ? "\n" + existingBody : ""}`;
  fs.writeFileSync(changelogFile, out);
}

/**
 * End-to-end bump: read current version, decide the next one from the commit
 * log, update the version file, and prepend a changelog entry.
 *
 * @param {object} opts
 * @param {string} opts.versionFile  path to VERSION / package.json
 * @param {string} opts.commitsFile  path to the commit log fixture/dump
 * @param {string} opts.changelogFile path to CHANGELOG.md
 * @param {string} opts.date         ISO date for the changelog header
 * @returns {{oldVersion:string,newVersion:string,bumpType:string,changelogEntry:string}}
 */
export function bump({ versionFile, commitsFile, changelogFile, date }) {
  const oldVersion = readVersion(versionFile);

  let log;
  try {
    log = fs.readFileSync(commitsFile, "utf8");
  } catch {
    throw new Error(`Commit log not found: ${commitsFile}`);
  }

  const commits = parseCommits(log);
  const bumpType = determineBump(commits);
  if (!bumpType) {
    throw new Error(
      "No conventional commits found that warrant a version bump " +
        "(need at least one feat, fix, or breaking change)."
    );
  }

  const newVersion = bumpVersion(oldVersion, bumpType);
  const changelogEntry = generateChangelog(newVersion, commits, date);

  writeVersion(versionFile, newVersion);
  prependChangelog(changelogFile, changelogEntry);

  return { oldVersion, newVersion, bumpType, changelogEntry };
}

File: results/2026-06-30_191904/tasks/11-semantic-version-bumper/default-sonnet5-low/generated-code/bumper.py
#!/usr/bin/env python3
"""
Semantic version bumper.

Reads a version from a version file (package.json or a plain VERSION file),
inspects a list of conventional-commit messages to decide whether the next
release is a major/minor/patch bump, writes the new version back to the
version file, appends a changelog entry, and prints the new version.

Conventional commit rules applied (in priority order):
  - "type!:" prefix or a "BREAKING CHANGE:" footer -> major
  - "feat:" prefix                                  -> minor
  - "fix:"  prefix                                   -> patch
  - anything else (docs, chore, style, refactor...)  -> ignored
"""
import argparse
import json
import re
import sys
from pathlib import Path


class BumperError(Exception):
    """Raised for any user-facing error (bad input, missing file, etc.)."""


_VERSION_RE = re.compile(r"^(\d+)\.(\d+)\.(\d+)$")
_COMMIT_HEADER_RE = re.compile(r"^(?P<type>\w+)(?P<breaking>!)?:\s*(?P<subject>.+)$")


def parse_version(version_str):
    """Parse a 'X.Y.Z' string into an (int, int, int) tuple."""
    match = _VERSION_RE.match(version_str.strip())
    if not match:
        raise BumperError(
            f"Invalid semantic version string: {version_str!r} "
            "(expected format MAJOR.MINOR.PATCH)"
        )
    return tuple(int(part) for part in match.groups())


def _split_commits(commits):
    """Split a list of raw commit blobs into individual header+body commits.

    Each element of `commits` may itself contain multiple lines (subject +
    body), so we treat each list element as one commit message.
    """
    return [c for c in commits if c.strip()]


def determine_bump(commits):
    """Inspect commit messages and return 'major', 'minor', 'patch', or None."""
    commits = _split_commits(commits)
    found = set()
    for commit in commits:
        lines = commit.splitlines()
        header = lines[0] if lines else ""
        body = "\n".join(lines[1:])

        if "BREAKING CHANGE:" in commit or "BREAKING-CHANGE:" in commit:
            found.add("major")
            continue

        match = _COMMIT_HEADER_RE.match(header.strip())
        if not match:
            continue

        if match.group("breaking"):
            found.add("major")
        elif match.group("type") == "feat":
            found.add("minor")
        elif match.group("type") == "fix":
            found.add("patch")

    if "major" in found:
        return "major"
    if "minor" in found:
        return "minor"
    if "patch" in found:
        return "patch"
    return None


def bump_version(version_tuple, bump_type):
    """Apply a bump_type ('major'/'minor'/'patch') to a version tuple."""
    major, minor, patch = version_tuple
    if bump_type == "major":
        return (major + 1, 0, 0)
    if bump_type == "minor":
        return (major, minor + 1, 0)
    if bump_type == "patch":
        return (major, minor, patch + 1)
    raise BumperError(f"Unknown bump type: {bump_type!r}")


def read_version_file(path):
    """Return the version string found in a package.json or plain VERSION file."""
    path = Path(path)
    if not path.exists():
        raise BumperError(f"Version file not found: {path}")

    if path.suffix == ".json":
        try:
            data = json.loads(path.read_text())
        except json.JSONDecodeError as exc:
            raise BumperError(f"Invalid JSON in {path}: {exc}") from exc
        if "version" not in data:
            raise BumperError(f"No 'version' key found in {path}")
        return data["version"]

    return path.read_text().strip()


def write_version_file(path, new_version):
    """Write new_version into a package.json (preserving other keys) or VERSION file."""
    path = Path(path)
    if path.suffix == ".json":
        data = json.loads(path.read_text())
        data["version"] = new_version
        path.write_text(json.dumps(data, indent=2) + "\n")
    else:
        path.write_text(new_version + "\n")


def generate_changelog_entry(new_version, commits):
    """Build a Markdown changelog entry grouping commits by conventional type."""
    commits = _split_commits(commits)
    breaking, features, fixes, other = [], [], [], []

    for commit in commits:
        lines = commit.splitlines()
        header = lines[0] if lines else ""
        match = _COMMIT_HEADER_RE.match(header.strip())
        is_breaking = "BREAKING CHANGE:" in commit or "BREAKING-CHANGE:" in commit

        if not match:
            continue

        subject = match.group("subject")
        if match.group("breaking") or is_breaking:
            breaking.append(subject)
        elif match.group("type") == "feat":
            features.append(subject)
        elif match.group("type") == "fix":
            fixes.append(subject)
        else:
            other.append(subject)

    lines = [f"## {new_version}", ""]
    if breaking:
        lines += ["### Breaking Changes", ""] + [f"- {s}" for s in breaking] + [""]
    if features:
        lines += ["### Features", ""] + [f"- {s}" for s in features] + [""]
    if fixes:
        lines += ["### Fixes", ""] + [f"- {s}" for s in fixes] + [""]
    return "\n".join(lines).rstrip() + "\n"


def _read_commits_file(path):
    """Read a commit-log fixture file, one commit message per line (blank-line separated)."""
    path = Path(path)
    if not path.exists():
        raise BumperError(f"Commits file not found: {path}")
    raw = path.read_text()
    # Commits are separated by blank lines; each block is one commit message
    # (which may itself span multiple lines, e.g. a BREAKING CHANGE footer).
    blocks = re.split(r"\n\s*\n", raw.strip())
    return [b for b in blocks if b.strip()]


def run(version_file, commits_file, changelog_file):
    """Orchestrate the full bump: read version, decide bump, write outputs."""
    version_file = Path(version_file)
    changelog_file = Path(changelog_file)

    current_version = read_version_file(version_file)
    current_tuple = parse_version(current_version)

    commits = _read_commits_file(commits_file)
    bump_type = determine_bump(commits)
    if bump_type is None:
        raise BumperError(
            "No version-bumping commits found (need at least one feat/fix/"
            "breaking-change commit)"
        )

    new_tuple = bump_version(current_tuple, bump_type)
    new_version = ".".join(str(p) for p in new_tuple)

    write_version_file(version_file, new_version)

    entry = generate_changelog_entry(new_version, commits)
    existing = changelog_file.read_text() if changelog_file.exists() else ""
    changelog_file.write_text(entry + ("\n" + existing if existing else ""))

    return {
        "old_version": current_version,
        "new_version": new_version,
        "bump_type": bump_type,
    }


def main(argv=None):
    parser = argparse.ArgumentParser(description="Semantic version bumper")
    parser.add_argument("--version-file", required=True)
    parser.add_argument("--commits-file", required=True)
    parser.add_argument("--changelog-file", required=True)
    args = parser.parse_args(argv)

    try:
        result = run(args.version_file, args.commits_file, args.changelog_file)
    except BumperError as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        return 1

    print(f"OLD_VERSION={result['old_version']}")
    print(f"NEW_VERSION={result['new_version']}")
    print(f"BUMP_TYPE={result['bump_type']}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
